feat: Martial Arts Expansion

acssimpson opened the pull request:
## Purpose of change (The Why)

The current martial arts system is very clunky/coarse in its approximation of a state machine, while also being quite limited in what it allows martial arts to do.  It should be updated to fix both of these problems so we can make existing stuff cleaner, and make new stuff cooler.

(TODO: Look for related issues and add them. I'm sure there are plenty)

## Describe the solution (The How)

This PR aims to greatly expand the JSON options available when defining a Martial Art, both to clean up/streamline existing logic, and to allow for more interesting/diverse patterns & mechanics. Most of the changes I am proposing either have an obvious intended use, or are an expansion of existing functionality.

The full list of planned changes can be found on [this spreadsheet](https://docs.google.com/spreadsheets/d/1EFPpdK1kwsvKgH3WNkspQf6E1IR049kWlzQkA5HPFk0/edit?gid=0#gid=0), but I will be attaching gifs and examples of each field below.

Also, TBD but I would really like to clean up the in-game martial arts description. It is VERY difficult to parse, and now seems like a good time since I will have to add entries for all of these field anyway.

## Describe alternatives you've considered

While limited, I think the martial arts system is also relatively *focused* right now, and I've tried to preserve that by not expanding it too much. Changes I considered but decided against include:
- Refactoring existing structures to be more generalized
          `"req_buffs": [ "buff_example_onhit" ]` -> `"req_effects": [ "buff_example_onhit", "grabbed" ]`
          `"downed_target": true` -> `"req_effects_target": [ "downed" ]`
- Building in cross pollination of other combat systems, like allowing techniques to apply any string-specified effect, or allowing them to trigger spells/enchants like the `on_hit_you` effects weapons can trigger.
- Anything that seemed like it would overcomplicate or add too much boiler plate JSON that might makes it more difficult to understand/work with the system.

## Testing

All of these effects will be tested. Gifs and JSON examples incoming.

## Additional context
Ideally this should not *remove* any functionality, from the game. There is an implied follow up task where existing clunky martial arts are rewritten using better logic enabled by this pr, but that is not included.

## Checklist

### Mandatory

- [x] I wrote the PR title in [conventional commit format](https://docs.cataclysmbn.org/en/contribute/changelog_guidelines/).
- [ ] I ran the [code formatter](https://docs.cataclysmbn.org/en/contribute/contributing/#code-style).
- [ ] I linked any relevant issues using [github keyword syntax](https://docs.cataclysmbn.org/en/contribute/contributing/#pull-request-notes) like `closes #1234` in [Summary of the PR](#why-should-this-pr-be-merged) so it can be closed automatically.
- [x] I've [committed my changes to new branch that isn't `main`](https://docs.cataclysmbn.org/en/contribute/contributing/#make-your-changes) so it won't cause conflict when updating `main` branch later.

### Optional

- [x] This is a C++ PR that modifies JSON loading or behavior.
  - [ ] I have documented the changes in the appropriate location in the `doc/` folder.
  - [ ] If documentation for this feature does not exist, please write it or at least note its lack in PR description.
  - [ ] New localizable fields need to be added to the `lang/bn_extract_json_strings.sh` script if it does not support them yet.
  - [ ] If applicable, add checks on game load that would validate the loaded data.
-->

acssimpson committed:
Update .gitignore

acssimpson committed:
Initial commit with martial arts changes.

Initial commit. Also, Added knockback_follow_full, strictly_running, consume_buffs (which isn't fully functional yet), annd switch_pos (also renamed side_switch to switch_side). Started work on reach attacks.

autofix-ci[bot] committed:
style(autofix.ci): automated formatting

acssimpson committed:
Revert "Update .gitignore"

This reverts commit 5b6f75d93f36b67f56f716519dbef61868bfa722.

acssimpson committed:
Merge branch 'MA-Changes' of https://github.com/acssimpson/Cataclysm-BN into MA-Changes

acssimpson committed:
Implemented many new JSON fields for martial arts, added new JSON intr. method

Added:
req_running
req_half_wall
switch_side_self
switch_pos
reach_tec
reach_ok
pull_target
pull_self
consumes_buffs
req_buffs
req_target_effects

Updated Names for:
side_switch -> switch_side_target
stealthy -> quiet_movement
quiet -> quiet_attacks
powerful_knockback -> knockback_type
knockback_follow -> knockback_follow_type

acssimpson committed:
Merge branch 'main' into MA-Changes

autofix-ci[bot] committed:
style(autofix.ci): automated formatting

Coolthulhu commented on src/martialarts.cpp:
```diff
@@ -534,10 +599,24 @@ bool ma_requirements::is_valid_character( const Character &u ) const
     }
 
-    if( wall_adjacent && !get_map().is_wall_adjacent( u.pos() ) ) {
+    if( req_running && !u.movement_mode_is( CMM_RUN ) ) {
```
Mixed feelings on running requirement.
On one hand, it allows mode switching, which is pretty powerful - same technique could switch between, say, offensive or defensive, charge forward or stay rooted etc.
On the other hand, it allows somewhat clunky, tedious mode switching.

I'm not saying no, it's just that there may be a better way to do mode switching - something that doesn't tie it to an unrelated mode option. Tying it to movement mode feels like a workaround (for lack of free keybinds) rather than "the right thing".
Still, if no better solution is found, this is good enough.

Coolthulhu commented on src/melee.cpp:
```diff
@@ -1451,47 +1463,97 @@ void Character::perform_technique( const ma_technique &technique, Creature &t, d
+                t.setpos( dest );
+            } else {
+                if( in_vehicle ) { //Don't unboard unless it's the player, and unless there is actually a place to swap to.
+                    get_map().unboard_vehicle( pos() );
```
This will need robust testing, boarding/unboarding code is fragile as fuck when combined with dragged vehicles.
Be sure to test at least the following cases:
* Player is dragging a cart, zombie stands on cart, player swaps with zombie, player moves off cart
* As above, but player is standing on cart2

Coolthulhu commented on src/melee.cpp:
```diff
@@ -1451,47 +1463,97 @@ void Character::perform_technique( const ma_technique &technique, Creature &t, d
     }
 
-    if( technique.stun_dur > 0 && !technique.powerful_knockback ) {
+    if( technique.stun_dur > 0 && technique.knockback_type.empty() ) {
```
Using `std::optional` would be better than testing for empty string.

Coolthulhu commented on src/melee.cpp:
```diff
@@ -1580,6 +1646,12 @@ void Character::perform_technique( const ma_technique &technique, Creature &t, d
+
+    for( const mabuff_id &consume_id : technique.reqs.consume_buffs ) {
+        if( has_mabuff( consume_id ) ) {
+            remove_effect( efftype_id( std::string( "mabuff:" ) + consume_id.str() ) );
```
Evil string manipulation.
In order to limit the evilness, move that to load/finalize time. This will enable verification if the `mabuff` exists on load time, rather than only during runtime.

acssimpson committed:
Implemented consumed_buffs, added new effect removal methods, other stuff

-consumed_buffs and required_buffs are now vector<pair<mabuff_id, int>>
-added new effect removal methods that allow for effects to have part of the intensity reduced externally. Not happy with the location of these methods and am going to change it.
-added more safety rails for existing fields that will soon be deprecated.

acssimpson committed:
Merge branch 'MA-Changes' of https://github.com/acssimpson/Cataclysm-BN into MA-Changes

autofix-ci[bot] committed:
style(autofix.ci): automated formatting

autofix-ci[bot] committed:
style(autofix.ci): automated formatting (attempt 2/3)

acssimpson committed:
Throwing techniques, sneak attack, prevented_by buffs, required_buffs, stack logic, and more!

-Added support for throwing techniques, including relevant flags.
-Added new JSON fields to improve MA state machine.
    req_buffs -> required_buffs
    consumed_buffs
    prevented_by_buffs

required_buffs and consumed_buffs have both been changed to pair vectors and now take into account desired stacks of a given buff.

Added sneak attack

Improved logic for catching old json fields to hopefully make backwards compatability a non-issue.

acssimpson committed:
Merge branch 'MA-Changes' of https://github.com/acssimpson/Cataclysm-BN into MA-Changes

autofix-ci[bot] committed:
style(autofix.ci): automated formatting
